src/rapidhtml/tags.py

```diff
@@ -126,6 +142,99 @@ def render(self):
 
         return ret_html
 
+    def select(
+        self,
+        select_tag: typing.Type["BaseTag"] | str,
+        *,
+        recurse: bool = False,
+        pop: bool = False,
+        first_match: bool = False,
+    ) -> list["BaseTag"]:
+        """
+        Selects and returns a list of BaseTag objects that match the given tag name or BaseTag object.
+
+        Args:
+            tag (Type[BaseTag] or str): The uninstantiated subclass of BaseTag or the tag name to match against.
+            recurse (bool, optional): If True, recursively searches for matching tags within nested BaseTag objects. Defaults to False.
+            pop (bool, optional): If True, removes the selected tags from the current object. Defaults to False.
+            first_match (bool, optional): If True, stops searching after finding the first match. Defaults to False.
+
+        Returns:
+            list[BaseTag]: A list of BaseTag objects that match the given tag name or BaseTag object.
+
+        Raises:
+            KeyError: If no matching tags are found and pop is True.
+            ValueError: If an instantiated object is passed as the tag parameter.
+            TypeError: If the tag parameter is of an unexpected type.
+
+        """
+        if isinstance(select_tag, BaseTag):
+            raise ValueError("Cannot pass instantiated object to select method!")
+        elif inspect.isclass(select_tag) and issubclass(select_tag, BaseTag):
+            select_tag_name = select_tag.__name__.lower()
+        elif isinstance(select_tag, str):
+            select_tag_name = select_tag
+        else:
+            raise TypeError(f"Unexpected tag type {type(select_tag)}. {select_tag}")
+
+        ret_tags: list["BaseTag"] = []
+        self_matches: list[tuple[int, "BaseTag"]] = []
+
+        for i, tag in enumerate(self.tags):
+            if not isinstance(tag, BaseTag):
+                continue
+            tag_name = tag.tag_name
+            if tag_name == select_tag_name or tag_name == f"htmltag{select_tag_name}":
+                self_matches.append((i, tag))
+                if first_match:
+                    break
+            elif recurse and isinstance(tag, BaseTag):
+                sub_ret_tags = tag.select(
```

Comment:
It's an edge case, but if we had tags of the same type nested inside each other (I can only really think of `div` that would do this), I don't think this logic would get the parent?

Reply:
Sir, I may need you to draw out this edge case a bit more. I am not following you on this

Reply:
So, I think my initial review was a bit off (I hadn't actually tested it yet), but looking at it now I think it's still a bit off

Say I have this:
```py
from rapidhtml.tags import *

html = Html(
    Div(
        Div(
            Div(
                id="3"
            ),
            id="2"
        ),
        id="1"
    )
)```
If I do `html.select(Div)` it seems I'll only get `[<rapidhtml.tags.Div at 0x7f048e55f730>]` which is the `Div` with `id="1"`

So tags nested inside other tags of the same type don't get selected. Though I can't think of any other tag that gets nested inside itself aside from `Div`, so it's a very low priority bug (who's trying to get all the `Div`s anyways?

Reply:
Ah I see what you are saying!

I thought about selecting tags based on the value of an attr, but I left it out for now, but I see this as an absolute enhancement on top of this that we should do!

Currently, if you knew the structure of the page, and wanted let's say Div 2, you'd have to do:

```python3
html.select(Div, first_match=True)[0].select(Div)
```

When you are selecting Div with Id1, it is returning the reference to the object, so you can subselect within it!